Refactor CompositeRolesStore for separation of concerns

ywangd opened the pull request:
The process of role resolving is to build the Role object from an
Authentication object. The high level steps of this process is as the follows:

1. Locate the role reference for the Authentication, e.g. for regular user,
this means a collection of role names.

2. Retrieve the role descriptors for the role reference, e.g. search the
security index to get the role descriptors for the role name.  

3. Build the Role object based on the role descriptors.  There are also special
cases in the above process. For example, API keys do not have role names, but
two byteReference representing the role descriptors. API keys also have a
nested Role structure for limiting the key's actual privileges based on the
owner's.

Today, this process is managed by a single CompositeRolesStore class and the
steps are not clearly separated. This PR improves the situation by introducing
a new RoleReferenceResolver class that is responsible for turning roleReference
into role descriptors. CompositeRolesStore is now only responsible for buiding
the Role from the descriptors.

The RoleReference is also a new concept introduced by this PR, along with
AuthenticationContext and Subject. They technically belong to the issue of
revisiting Authentication class (#80117). But they are needed here to faciliate
the changes. Their usage will be expanded once we start working on #80117.

A Subject knows how to return a list of RoleReference and the final Role is the
intersection of all the RoleReference. This was a concept for API keys. It is
now generalised in this PR in the light on potential expanded usage of
limitedBy roles.

Relates: #80117

ywangd commented:
@elasticmachine run elasticsearch-ci/part-2

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:
```diff
@@ -151,239 +132,156 @@ public void providersChanged() {
-    // for testing
-    Role getAsyncSearchUserRole() {
-        return asyncSearchUserRole;
+        final List<RoleReference> roleReferences = subject.getRoleReferences(anonymousUser);
+        // TODO: Two levels of nesting can be relaxed in future
+        assert roleReferences.size() <= 2 : "only support up to one level of limiting";
+
+        buildRoleFromRoleReference(roleReferences.get(0), ActionListener.wrap(role -> {
+            if (roleReferences.size() == 1) {
+                roleActionListener.onResponse(role);
+            } else {
+                buildRoleFromRoleReference(
+                    roleReferences.get(1),
+                    ActionListener.wrap(
+                        limitedByRole -> roleActionListener.onResponse(LimitedRole.createLimitedRole(role, limitedByRole)),
+                        roleActionListener::onFailure
+                    )
+                );
+            }
+
+        }, roleActionListener::onFailure));
```
We only support 2 levels of role nesting right now because how `LimitedRole` works. These code can be generalised to support multiple levels once we re-work how Role nesting is represented (likely need an interface for `Role`)

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DefaultRoleReferenceResolver.java:
```diff
@@ -0,0 +1,252 @@
+            logger.trace(
+                () -> new ParameterizedMessage("Building role from descriptors [{}] for role names [{}]", effectiveDescriptors, roleNames)
+            );
+            // TODO: why not populate negativeLookupCache here with missing roles?
```
The existing code populates the negativeLookupCache [after](https://github.com/elastic/elasticsearch/blob/85d12e5b5b45021a91078b2fce3030b36ccc8f0b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L386-L388) the `Role` gets successfully built.

I chose to retain this behaviour. But I wonder why it cannot be populated earlier here straight after roleDescriptor retrevial since we already know what roles are missing at this point?

tvernum replied:
I don't think there's any reason for that. I think it just ended up there after introducing the RoleRetrievalResult

ywangd replied:
I am going to leave it as is for now because:
1) It retains the current behaviour 
2) I'd like to move the effective role descriptor checking (licensing) from the Resolver back to CompositeRolesStore. My thinking is the the Resolver should just resolve the role descriptors and the caller will handle both the license check and both negative and positive caching. But this change has a decent size and potentially delay the process of this PR. So I'd like to handle it in a follow-up (I added a task in https://github.com/elastic/elasticsearch/issues/80117).

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:
```diff
@@ -0,0 +1,199 @@
+        this.version = version;
+        this.user = user;
+        this.realm = realm;
+        // Realm can be null for run-as user if it does not exist. Pretend it is a user and it will be rejected later in authorization
+        // This is to be consistent with existing behaviour.
+        if (realm == null) {
+            this.type = Type.USER;
```
This is a bit odd, but it is to retain the existing error reporting behaviour. We could argue that the existing error message is misleading anyway because it says something like
> action [...] is unauthorized for user [elastic] run as [foo] with roles [], this action is granted by the cluster privileges [...]

which feels like the fix is to grant some roles to the user `foo` but the actual problem is that user `foo` does not exist. I think we should fix it. But it should be a separate PR.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:
```diff
@@ -0,0 +1,199 @@
+        return metadata;
+    }
+
+    /**
+     * Return a List of RoleReferences that represents role definitions associated to the subject.
+     * The final role of this subject should be the intersection of all role references in the list.
+     */
+    public List<RoleReference> getRoleReferences(@Nullable AnonymousUser anonymousUser) {
```
In the eariler version, I wrapped the List inside a dedicate class. But decided later to just return the list directly. Because (1) I didn't come up with other useful field or method for that class; (2) the usage of the list is quite simple, it must be intersected and order not does not even matter.

tvernum replied:
I prefer a separate class, but we can handle it as a separate discussion.
My main reason is that I don't think a `List` means anything without a lot of context, but `RoleIntersection` does.

However, if we're going to do a separate PR to handle `LimitedRole` etc, we can bundle up a series of changes there.

ywangd replied:
Makes sense to me. I raised https://github.com/elastic/elasticsearch/issues/81192 and let's sort it out there.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java:
```diff
@@ -0,0 +1,163 @@
+
+        @Override
+        public RoleKey cacheKey() {
+            // Since api key id is unique, it is sufficient and more correct to use it as the names
+            return new RoleKey(Set.of(apiKeyId), apiKeyId + roleKeySourceSuffix);
```
The existing code use API key role names as the first argument to `RoleKey' constructor. It's hard to get that information with the new structure and it is also unnecessary. Since API key ID is used as the source of `RoleKey`. It is already guaranteed that one entry for each single key. So I just use the API key ID again as the first argument.

tvernum replied:
Why do we include the api key id in the source field? Is that needed?

ywangd replied:
It is not needed. I just copied it from the existing code. But it was needed there because the names are not guaranteed to be unique. Since we now use the API Key ID as the names, it is no longer necesesary to have it in the source. I'll drop it.

ywangd commented:
@elasticmachine run elasticsearch-ci/part-2

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:
```diff
@@ -0,0 +1,199 @@
+            if (anonymousUser.roles().length == 0) {
+                throw new IllegalStateException("anonymous is only enabled when the anonymous user has roles");
+            }
+            allRoleNames = Stream.concat(Arrays.stream(user.roles()), Arrays.stream(anonymousUser.roles())).toArray(String[]::new);
```
```suggestion
            allRoleNames = ArrayUtils.concat(user.roles(), anonymousUser.roles());
```

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    private List<RoleReference> buildRoleReferencesForApiKey() {
+        if (false == version.onOrAfter(VERSION_API_KEY_ROLES_AS_BYTES)) {
```
I think this reads more cleanly:
```suggestion
        if (version.before(VERSION_API_KEY_ROLES_AS_BYTES)) {
```

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:
```diff
@@ -0,0 +1,199 @@
+            limitedByRoleDescriptorsBytes,
+            "apikey_limited_role"
+        );
+        if (roleDescriptorsBytes == null && limitedByRoleDescriptorsBytes == null) {
```
Should this be moved up 1 statement, before we try to use `limitedByRoleDescriptorsBytes` ?

ywangd replied:
Yes it should. Thanks for catching it.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleKey.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Objects;
+import java.util.Set;
+
+public final class RoleKey {
```
Javadoc please?

ywangd replied:
Added.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java:
```diff
@@ -0,0 +1,163 @@
+     * The key for caching purpose. Callers can use this value to determine whether it should skip
+     * resolving the role descriptors and subsequently building the role.
+     */
+    RoleKey cacheKey();
```
Do we explicitly want to call this a cache key? or simply an id?
Mostly I'm wondering if there's places where it would be useful to be able to describe a RoleReference (as an id), but maybe that's handled by `toString()`?

Calling it a cacheKey just feels artificially restrictive to me.

ywangd replied:
I changed it to `id`. Your suggestion makes sense to me. Also I think it is better to avoid using the word `cache` in `RoleReference` since in theory it does not know and should not care about caching.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java:
```diff
@@ -0,0 +1,163 @@
+        public RoleKey cacheKey() {
+            final String roleDescriptorsHash = MessageDigests.toHexString(
+                MessageDigests.digest(roleDescriptorsBytes, MessageDigests.sha256())
+            );
```
Should this key be memoized? If there's any chance we'd call the digest twice then that seems unnecessarily expensive.

ywangd replied:
It is only called once by `CompositeRolesStore` and the result `RoleKey` is then passed down to other methods. So I don't think it is necessary memorize. That said, since we agree to change this method name to `id`. This gives it an impression that it could be called in other places (in future). So I added a simple class field to hold the result. Didn't use the MemorizedSupplier or volatile since they feel like overkill at this point.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java:
```diff
@@ -0,0 +1,163 @@
+    /**
+     * Resolve concrete role descriptors for the roleReference.
+     */
+    void resolve(RoleReferenceResolver resolver, ActionListener<RolesRetrievalResult> listener);
```
All the implementations of this method just call another method on the `resolver`. Do we really want that indirection, or should we just remove this method and expect callers to know that they should consult `RoleReferenceResolver` ? (via docs)

ywangd replied:
I acutally quite like the indirection. It feels more OO because no `instanceof` is needed. It also feels more fluent from the caller side.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RolesRetrievalResult.java:
```diff
@@ -0,0 +1,52 @@
+public final class RolesRetrievalResult {
+
+    public static final RolesRetrievalResult EMPTY = new RolesRetrievalResult();
+    public static final RolesRetrievalResult SUPERUSER = new RolesRetrievalResult();
```
Can we merge the code for this and that `static` block somehow. It bothers me that `EMPTY` and `SUPERUSER` are the same at construction time.

Either:
```
public static final RolesRetrievalResult SUPERUSER;
static {
 SUPERUSER = new RolesRetrievalResult();
 SUPERUSER.addDescriptors(Set.of(ReservedRolesStore.SUPERUSER_ROLE_DESCRIPTOR));
}
```
or

```
     public static final RolesRetrievalResult SUPERUSER = superuserRolesRetrievalResult();
     private static RolesRetrievalResult superuserRolesRetrievalResult() { ... }
```
?
```

ywangd replied:
I took the first approach by merging them in the static block.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/User.java:
```diff
@@ -72,7 +72,7 @@ private User(
     /**
      * @return  The principal of this user - effectively serving as the
-     *          unique identity of of the user.
+     *          unique identity of the user.
```
```suggestion
     *          unique identity of the user (within a given realm).
```

tvernum reviewed (commented):
I've reviewed the `/main/` code for `/core/`, but I need to break for the day.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DefaultRoleReferenceResolver.java:
```diff
@@ -0,0 +1,252 @@
+        RoleReference.NamedRoleReference namedRoleReference,
+        ActionListener<RolesRetrievalResult> listener
+    ) {
+        final Set<String> roleNames = Set.copyOf(new HashSet<>(List.of(namedRoleReference.getRoleNames())));
```
Why the nested collections?
```suggestion
        final Set<String> roleNames = Set.of(namedRoleReference.getRoleNames());
```

ywangd replied:
Unfortunately, `namedRoleReference.getRoleNames()` can have duplicate names and `Set.of` does not like it. But the outermost `Set.copyOf` is not necessary (I just added it out of habit) and I can drop it if you prefer.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DefaultRoleReferenceResolver.java:
```diff
@@ -0,0 +1,252 @@
+            );
+            effectiveRoleDescriptorsConsumer.accept(Collections.unmodifiableCollection(effectiveDescriptors));
+            logger.trace(
+                () -> new ParameterizedMessage("Building role from descriptors [{}] for role names [{}]", effectiveDescriptors, roleNames)
```
It looks like we have 2 trace messages with essentially the same content here (with line 153 above)

ywangd replied:
Right. Not sure why we had the 2nd one. I removed it.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DefaultRoleReferenceResolver.java:
```diff
@@ -0,0 +1,252 @@
+import static java.util.function.Predicate.not;
+import static org.elasticsearch.xpack.core.security.SecurityField.DOCUMENT_LEVEL_SECURITY_FEATURE;
+
+public class DefaultRoleReferenceResolver implements RoleReferenceResolver {
```
```suggestion
public class RoleDescriptorStore implements RoleReferenceResolver {
```
? or something like that ?

ywangd replied:
I personally prefer the prefix naming style of implementation class. But it is not the convention used in this code base. So I changed it to `RoleDescriptorStore`

tvernum reviewed (commented):
I've now reviewed everything except tests.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:
```diff
@@ -151,239 +132,156 @@ public void providersChanged() {
+        // TODO: Two levels of nesting can be relaxed in future
+        assert roleReferences.size() <= 2 : "only support up to one level of limiting";
+
+        buildRoleFromRoleReference(roleReferences.get(0), ActionListener.wrap(role -> {
```
No it cannot and I don't think we should allow it. The `RoleReference` can already handle empty role names well. It's too much flexibility if the list is allowed to be empty as well which does not add any benefit, just more edge cases. I added an assertion for empty list.

ywangd commented:
This is ready for another look.

tvernum commented:
It looks like something has confused GitHub - it thinks you have 67 commits with 357 modified files.

<img width="772" alt="image" src="https://user-images.githubusercontent.com/2244393/144353476-703822ed-b8a4-4450-8435-c3277f289fd8.png">

What can we do to get this back into a reviewable state?

ywangd committed:
Refactor CompositeRolesStore for speration of concerns

The process of role resolving is to build the Role object from an
Authentication object. The high level steps of this process is as the follows:

1. Locate the role reference for the Authentication, e.g. for regular user,
this means a collection of role names.

2. Retrieve the role descriptors for the role reference, e.g. search the
security index to get the role descriptors for the role name.

3. Build the Role object based on the role descriptors.  There are also special
cases in the above process. For example, API keys do not have role names, but
two byteReference representing the role descriptors. API keys also have a
nested Role structure for limiting the key's actual privileges based on the
owner's.

Today, this process is managed by a single CompositeRolesStore class and the
steps are not clearly separated. This PR improves the situation by introducing
a new RoleReferenceResolver class that is responsible for turning roleReference
into role descriptors. CompositeRolesStore is now only responsible for buiding
the Role from the descriptors.

The RoleReference is also a new concept introduced by this PR, along with
AuthenticationContext and Subject. They technically belong to the issue of
revisiting Authentication class (#80117). But they are needed here to faciliate
the changes. Their usage will be expanded once we start working on #80117.

A Subject knows how to return a list of RoleReference and the final Role is the
intersection of all the RoleReference. This was a concept for API keys. It is
now generalised in this PR in the light on potential expanded usage of
limitedBy roles.

Relates: #80117

ywangd commented:
> What can we do to get this back into a reviewable state?

I force-pushed. It didn't preserve previous commit message but at least get the PR in a sane state. I hope that works with you. Thanks!

ywangd commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into revisit-authentication-class

tvernum commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/SubjectTests.java:
```diff
@@ -0,0 +1,120 @@
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.isA;
+
+public class SubjectTests extends ESTestCase {
```
I assume this is a result of moving things (including tests) around between classes, but it seems strange that this class only has tests for API Keys.

I would prefer there were more tests.

ywangd replied:
I added more tests to cover other subject types.

tvernum reviewed (approved):
LGTM, though the tests seem a little too narrow in scope.

ywangd committed:
address feedback to add more tests

ywangd committed:
Merge remote-tracking branch 'origin/master' into revisit-authentication-class

ywangd commented:
> LGTM, though the tests seem a little too narrow in scope.

Thanks Tim. I added more tests for both Subject and RoleReference in the latest commit https://github.com/elastic/elasticsearch/pull/80926/commits/31dc580e09699f8b19931de780a48c5cc7f90579
